File: src/HolzShots.Legacy/Drawing/Tools/Censor.vb
Imports System.Drawing.Drawing2D
Imports System.Drawing.Text
Imports System.Numerics
Imports HolzShots.Drawing.Tools.UI

Namespace Drawing.Tools
    Friend NotInheritable Class Censor
        Implements ITool(Of CensorSettings)

        Private _pointList As List(Of Point)

        Private _beginCoordinates As Vector2
        Public Property BeginCoordinates As Vector2 Implements ITool(Of CensorSettings).BeginCoordinates
            Get
                Return _beginCoordinates
            End Get
            Set(value As Vector2)
                _beginCoordinates = value
                _pointList = New List(Of Point) From {_beginCoordinates.ToPoint2D()}
            End Set
        End Property

        Public Property EndCoordinates As Vector2 Implements ITool(Of CensorSettings).EndCoordinates

        Public ReadOnly Property Cursor As Cursor Implements ITool(Of CensorSettings).Cursor
            Get
                Dim censorWidth = Math.Max(5, SettingsControl.Settings.Width)
                Dim bmp As New Bitmap(Convert.ToInt32(0.2 * censorWidth), censorWidth)
                bmp.MakeTransparent()
                Using g As Graphics = Graphics.FromImage(bmp)
                    g.Clear(Color.FromArgb(200, 255, 0, 0))
                End Using
                Cursor = New Cursor(bmp.GetHicon())
            End Get
        End Property

        Public ReadOnly Property ToolType As ShotEditorTool = ShotEditorTool.Censor Implements ITool(Of CensorSettings).ToolType

        Public ReadOnly Property SettingsControl As ISettingsControl(Of CensorSettings) Implements ITool(Of CensorSettings).SettingsControl

        Sub New()
            SettingsControl = New CensorSettingsControl(CensorSettings.Default)
            _pointList = New List(Of Point)
        End Sub

        Public Sub LoadInitialSettings() Implements ITool(Of CensorSettings).LoadInitialSettings
            If My.Settings.ZensursulaWidth > CensorSettings.MaximumWidth OrElse My.Settings.ZensursulaWidth < CensorSettings.MinimumWidth Then
                My.Settings.ZensursulaWidth = CensorSettings.Default.Width
                My.Settings.Save()
            End If

            With SettingsControl.Settings
                .Width = My.Settings.ZensursulaWidth
                .Color = My.Settings.ZensursulaColor
            End With
        End Sub

        Public Sub PersistSettings() Implements ITool(Of CensorSettings).PersistSettings
            With SettingsControl.Settings
                My.Settings.ZensursulaWidth = .Width
                My.Settings.ZensursulaColor = .Color
            End With
        End Sub

        Private Shared Function CreatePen(settings As CensorSettings) As Pen
            Return New Pen(Color.FromArgb(255, settings.Color), settings.Width) With {
                .LineJoin = LineJoin.Round
            }
        End Function

        Public Sub RenderFinalImage(ByRef rawImage As Image) Implements ITool(Of CensorSettings).RenderFinalImage
            _pointList.Add(EndCoordinates.ToPoint2D())
            Using g = Graphics.FromImage(rawImage)
                Using censorPen = CreatePen(SettingsControl.Settings)
                    With g
                        .SmoothingMode = SmoothingMode.AntiAlias
                        .TextRenderingHint = TextRenderingHint.AntiAlias
                        If _pointList.Count > 0 AndAlso (_pointList.Count - 1) Mod 3 = 0 Then
                            .DrawBeziers(censorPen, _pointList.ToArray())
                        Else
                            .DrawBeziers(censorPen, _pointList.Take(_pointList.Count - (_pointList.Count - 1) Mod 3).ToArray())
                        End If
                    End With
                End Using
            End Using
            _pointList.Clear()
        End Sub

        Public Sub RenderPreview(rawImage As Image, g As Graphics) Implements ITool(Of CensorSettings).RenderPreview
            _pointList.Add(EndCoordinates.ToPoint2D())
            g.SmoothingMode = SmoothingMode.AntiAlias
            g.TextRenderingHint = TextRenderingHint.AntiAlias

            If _pointList.Count <= 0 Then
                Return
            End If

            Using censorPen = CreatePen(SettingsControl.Settings)
                Dim bs As Byte() = New Byte(_pointList.Count - 1) {}
                bs(0) = CByte(PathPointType.Start)
                For a = 1 To _pointList.Count - 1
                    bs(a) = CByte(PathPointType.Line)
                    g.DrawPath(censorPen, New GraphicsPath(_pointList.ToArray, bs))
                Next
            End Using
        End Sub

        Protected Sub Dispose() Implements ITool(Of CensorSettings).Dispose
            ' Nothing to do here
        End Sub

        Public Sub MouseOnlyMoved(rawImage As Image, ByRef currentCursor As Cursor, e As MouseEventArgs) Implements ITool(Of CensorSettings).MouseOnlyMoved
            ' Nothing to do here
        End Sub
        Public Sub MouseClicked(rawImage As Image, e As Vector2, ByRef currentCursor As Cursor, trigger As Control) Implements ITool(Of CensorSettings).MouseClicked
            ' Nothing to do here
        End Sub
    End Class
End Namespace

File: src/HolzShots.LegacyNew/Drawing/Tools/Censor.cs
using System.Drawing.Drawing2D;
using System.Drawing.Text;
using System.Numerics;
using HolzShots.Drawing.Tools.UI;

namespace HolzShots.Drawing.Tools;

public class Censor : ITool<CensorSettings>
{
    private List<Point> _pointList = [];

    private Vector2 _beginCoordinates;
    public Vector2 BeginCoordinates
    {
        get => _beginCoordinates;
        set
        {
            _beginCoordinates = value;
            _pointList = [_beginCoordinates.ToPoint2D()];
        }
    }

    public Vector2 EndCoordinates { get; set; }

    public Cursor Cursor
    {
        get
        {
            var censorWidth = Math.Max(5, SettingsControl.Settings.Width);
            var bmp = new Bitmap(Convert.ToInt32(0.2 * censorWidth), censorWidth);
            bmp.MakeTransparent();
            using var g = Graphics.FromImage(bmp);
            g.Clear(Color.FromArgb(200, 255, 0, 0));
            return new Cursor(bmp.GetHicon());
        }
    }

    public ShotEditorTool ToolType { get; } = ShotEditorTool.Censor;

    public ISettingsControl<CensorSettings> SettingsControl { get; } = new CensorSettingsControl(CensorSettings.Default);

    public void LoadInitialSettings()
    {
        if (Properties.Settings.Default.CensorWidth > CensorSettings.MaximumWidth || Properties.Settings.Default.CensorWidth < CensorSettings.MinimumWidth)
        {
            Properties.Settings.Default.CensorWidth = CensorSettings.Default.Width;
            Properties.Settings.Default.Save();
        }
        var settings = SettingsControl.Settings;
        settings.Width = Properties.Settings.Default.CensorWidth;
        settings.Color = Properties.Settings.Default.CensorColor;
    }

    public void PersistSettings()
    {
        var settings = SettingsControl.Settings;
        Properties.Settings.Default.CensorWidth = settings.Width;
        Properties.Settings.Default.CensorColor = settings.Color;
    }

    private static Pen CreatePen(CensorSettings settings)
    {
        return new Pen(Color.FromArgb(255, settings.Color), settings.Width)
        {
            LineJoin = LineJoin.Round
        };
    }

    public void RenderFinalImage(ref Image rawImage)
    {
        _pointList.Add(EndCoordinates.ToPoint2D());
        using var g = Graphics.FromImage(rawImage);

        using var censorPen = CreatePen(SettingsControl.Settings);
        g.SmoothingMode = SmoothingMode.AntiAlias;
        g.TextRenderingHint = TextRenderingHint.AntiAlias;

        if (_pointList.Count > 0 && (_pointList.Count - 1) % 3 == 0)
            g.DrawBeziers(censorPen, _pointList.ToArray());
        else
            g.DrawBeziers(censorPen, _pointList.Take(_pointList.Count - (_pointList.Count - 1) % 3).ToArray());

        _pointList.Clear();
    }

    public void RenderPreview(Image rawImage, Graphics g)
    {
        _pointList.Add(EndCoordinates.ToPoint2D());

        g.SmoothingMode = SmoothingMode.AntiAlias;
        g.TextRenderingHint = TextRenderingHint.AntiAlias;

        if (_pointList.Count <= 0)
            return;

        using var censorPen = CreatePen(SettingsControl.Settings);

        var bs = new byte[_pointList.Count - 1 + 1];
        bs[0] = Convert.ToByte(PathPointType.Start);
        for (var a = 1; a <= _pointList.Count - 1; a++)
        {
            bs[a] = Convert.ToByte(PathPointType.Line);
            g.DrawPath(censorPen, new GraphicsPath(_pointList.ToArray(), bs));
        }
    }

    public void Dispose() { }
    public void MouseOnlyMoved(Image rawImage, ref Cursor currentCursor, MouseEventArgs e) { }
    public void MouseClicked(Image rawImage, Vector2 e, ref Cursor currentCursor, Control trigger) { }
}
